src/Sparrow.Server/ByteString.cs

```diff
@@ -1935,6 +1947,16 @@ private void ReleaseSegment(SegmentInformation segment)
                 Allocator.Free(segment.Memory);
             }
         }
+        
+        public void RegisterListener(INotifyAllocationFailure listener)
+        {
+            Debug.Assert(listener != null);
+
+            if (_allocationFailureListener is null)
+                _allocationFailureListener = listener;
+            else
+                AllocationFailedMultiListener += () => listener.OnAllocationFailure(this);
```

Comment:
Why not have a single field + an array? Why use this sort of indirection?

Reply:
Because the delegates get special treatment by the runtime. There are a few edge cases I couldn't make to work with an array (that was the initial attempt), so I ended up going this route instead.